File: 515. Find Largest Value in Each Tree Row.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function(root) {
    const stack = [root];
    const result = [];
    let count = 1;
    let current = 0;
    let max = -Infinity;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            max = Math.max(max, node.val);
            current++;
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            if (current === count) {
                result.push(max);
                max = -Infinity;
                count = stack.length;
                current = 0;
            }
        }
    }
    return result;
};

File: Golang/Tree/515.find-largest-value-in-each-tree-row.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func largestValues(root *TreeNode) []int {
	ret := []int{}
	if root != nil {
		stack := []*TreeNode{root}
		for len(stack) > 0 {
			maxNum := math.MinInt32
			children := []*TreeNode{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				if maxNum < node.Val {
					maxNum = node.Val
				}
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			ret = append(ret, maxNum)
			stack = children
		}
	}
	return ret
}
